pool.js

```diff
@@ -337,9 +423,54 @@ Pool.prototype.remove_endpoint = function (host_port) {
     }
 };
 
+// adjusts the pool size within max_pool_size boundary if necessary
+// after we add/remove a discrete element from the pool, this adjustment must
+// carefully select a random element to remove/add so that connections are distributed
+// evenly across the cluster
+Pool.prototype.adjust_pool_size = function adjust_pool_size() {
+    // length can only be out of sync with all hostports if there's a max_pool_size
+    if (!this.max_pool_size) {
+        return;
+    }
+    var expected_size = Math.min(this.max_pool_size, this.all_hostports.length);
+
+    if (expected_size === this.length) {
+        return;
+    }
+
+    while (this.length !== expected_size) {
+        var idx;
+        if (this.length < expected_size) {
+            // add a random host from the all host_ports
+            var unused = this.unused_hostports();
+            idx = Math.floor(Math.random() * unused.length);
+            var randHostPort = unused[idx];
+            if (!this.endpoints_by_name[randHostPort]) {
+                this.add_pool_endpoint(randHostPort);
+            }
+        } else if (this.length > expected_size) {
+            idx = Math.floor(Math.random() * this.length);
+            var randEndpoint = this.endpoints[idx];
+            this.remove_pool_endpoint(randEndpoint.ip + ':' + randEndpoint.port);
```

Comment:
no.  you can see vaguely this in the PoolEndpoint constructor above,  it takes  ip+port.